ingest new Propolis VM creation API

gjcolombo opened the pull request:
Update Omicron to use the new Propolis VM creation API defined in oxidecomputer/propolis#813 and oxidecomputer/propolis#816. This API requires clients to pass instance specs to create new VMs and component replacement lists to migrate existing VMs. Construct these in sled agent for now; in the future this logic can move to Nexus and become part of a robust virtual platform abstraction. For now the goal is just to keep everything working for existing VMs while adapting to the new Propolis API.

Slightly adjust the sled agent instance APIs so that Nexus specifies disks and boot orderings using sled-agent-defined types and not re-exported Propolis types.

Finally, adapt Nexus to the fact that Crucible's `VolumeConstructionRequest` and `CrucibleOpts` types no longer appear in Propolis's generated client (and so don't appear in sled agent's client either). Instead, the `propolis-client` crate re-exports these types directly from its `crucible-client-types` dependency. For the most part, this involves updating `use` directives and storing `SocketAddr`s in their natively typed form instead of converting them to and from strings.

Tests: cargo nextest, plus ad hoc testing in a dev cluster as described in the PR comments.

gjcolombo commented on sled-agent/types/src/instance.rs:
```diff
@@ -153,6 +185,7 @@ pub struct VmmUnregisterResponse {
 #[derive(Copy, Clone, Debug, Deserialize, Serialize, JsonSchema)]
 pub struct InstanceMigrationTargetParams {
     /// The Propolis ID of the migration source.
+    // TODO(gjc) try to get rid of this, migration targets don't use it anymore
```
Need to clean this up still.

gjcolombo commented on sled-agent/src/instance.rs:
```diff
@@ -854,10 +854,79 @@ impl InstanceRunner {
+                    id: disk.disk_id,
+                    device: NvmeDisk {
+                        backend_id: SpecKey::Uuid(disk.disk_id),
+                        pci_path: PciPath::new(0, disk.slot + 0x10, 0)?,
```
This line (and the one below for NICs) could use an explanation as to why we're adding magic numbers to the slot number here.

gjcolombo reviewed (commented):
Making notes of a couple of things I saw while writing the PR description.

I also need to check that disk snapshots against a running instance work as intended, since that API also changed (I've tested this on the Propolis side but need to make sure it works here too).

gjcolombo commented:
> I also need to check that disk snapshots against a running instance work as intended, since that API also changed (I've tested this on the Propolis side but need to make sure it works here too).

I did this by spinning up a dev cluster, creating an instance and logging into it, touching some files there, and taking a snapshot of the image's disk while it was still running. The Propolis logs showed it servicing the snapshot request, and afterward I had a usable snapshot that I could use to create a new image/disk/instance that contains the files I'd touched pre-snapshot.

gjcolombo commented:
I've also tested the following:

- Disk snapshot: started an instance with a disk attached, took a snapshot, verified that the Propolis logs reflected that the disk snapshot API was called and that it returned success
- VCR replacement: expunged one of the simulated disks backing the instance's disk, verified that the Propolis logs showed a VCR replacement and a subsequent attempt to repair the disk

Looking into the integration test failures now.

gjcolombo commented:
The most recent integration test failure (in `test_disk_move_between_instances`) appears to be a preexisting race condition in simulated sled agent that's not related to these changes. I filed #7282 to track it and will rerun this job.

jmpesp commented on sled-agent/src/instance.rs:
```diff
@@ -854,10 +854,81 @@ impl InstanceRunner {
+                    id: disk.disk_id,
+                    device: NvmeDisk {
+                        backend_id: SpecKey::Uuid(disk.disk_id),
+                        // Add 16 to the disk slot number to match Propolis's
+                        // previous interpretation of disk slot parameters.
```
non-blocking wondering: will this number change in the future?

hawkw replied:
and, is there a thing to reference?

jmpesp commented on sled-agent/src/instance.rs:
```diff
@@ -867,64 +938,188 @@ impl InstanceRunner {
+                    device: VirtioNic {
+                        backend_id: SpecKey::Uuid(nic.id),
+                        interface_id: nic.id,
+                        // Add 8 to the NIC slot number to match Propolis's
+                        // previous interpretation of NIC slot parameters.
```
same question here

jmpesp commented on sled-agent/src/instance.rs:
```diff
@@ -867,64 +938,188 @@ impl InstanceRunner {
+                        backend_id: SpecKey::Name(
+                            "cloud-init-backend".to_string(),
+                        ),
+                        pci_path: PciPath::new(0, 0x18, 0).unwrap(),
```
0x18 is a magic number here

jmpesp commented on sled-agent/types/src/instance.rs:
```diff
@@ -47,6 +47,39 @@ pub struct InstanceEnsureBody {
+    //
+    // This is marked as `NoDebug` because the VCR contains the volume's
+    // encryption keys.
+    pub vcr_json: NoDebug<String>,
```
my assumption here is `NoDebug` means that it will be serialized/deserialized, just not printed to logs via the Debug derive, correct?

hawkw replied:
Yup.

jmpesp commented on sled-agent/types/src/instance.rs:
```diff
@@ -152,9 +184,6 @@ pub struct VmmUnregisterResponse {
 #[derive(Copy, Clone, Debug, Deserialize, Serialize, JsonSchema)]
 pub struct InstanceMigrationTargetParams {
-    /// The Propolis ID of the migration source.
-    pub src_propolis_id: Uuid,
```
It seems like a mistake to lose this, from the perspective of wishing we _had_ sent the expected UUID of the Downstairs we're connecting to from the Upstairs, and the bugs _not having it_ has caused.

jmpesp replied:
https://github.com/oxidecomputer/omicron/issues/6353 is a manifestation of the bug I'm talking about: an Upstairs constructed by a Pantry panicked when one of the regions it was connecting to did not match the others in the region set. The only information the Upstairs receives about Downstairs is SocketAddrs to connect to:

```
targets: [
  addr1, // downstairs A
  addr2, // downstairs B
  addr3  // downstairs C
]
```

When the Crucible agent creates and destroys Downstairs, it will aggressively reuse what ports are assigned to new Downstairs. If Nexus destroys Downstairs C, then creates a new Downstairs D (without doing anything else with the Agent), then Downstairs D will bind to the same port that Downstairs C did. If that happens then the targets array that the Upstairs was sent is instantly invalid, and constructing it will either 1) lead to a panic because the Downstairs D region is a different size than A and B, or 2) _not_ panic because Downstairs D is the same size as A and B, but it will be blank, leading to other issues.

Without any additional information in the targets array, the Upstairs has to panic: it would be invalid to continue and operate on regions with mismatched sizes. Adding the expected UUID would give the Upstairs the option to bail instead:

```
targets: [
  { uuidA, addr1 },
  { uuidB, addr2 },
  { uuidC, addr3 }
]
```

During the negotiation with the downstairs it would be able to compare the expected UUID in the targets array with what the Downstairs it's talking to is reporting, and fail to connect instead of panicking.

jmpesp commented on sled-agent/src/instance.rs:
```diff
@@ -854,10 +854,81 @@ impl InstanceRunner {
+            .map(|disk| -> Result<DiskComponents, std::io::Error> {
+                // NVMe disk serial numbers are supposed to be left-justified
+                // and padded with spaces (see version 1.2.1 of the spec,
+                // section 1.5). However, some versions of Propolis padded
+                // serial numbers with zeroes instead. Disk serial numbers can
```
will we always do this? or could this be a part of some versioning that will eventually be done away with in a V2?

hawkw replied:
Yeah, we would like to fix this eventually. I believe oxidecomputer/propolis#790 and perhaps also oxidecomputer/propolis#789 are the relevant Propolis issues; perhaps this ought to be documented someplace here?

jmpesp commented on sled-agent/src/instance.rs:
```diff
@@ -867,64 +938,188 @@ impl InstanceRunner {
+                .expect("migration ensure request entails migration in")
+                .migration_id;
+
+            let mut replace_components: std::collections::HashMap<_, _> = disks
```
Reading this section is a little strange, because it doesn't look like any kind of replacement is occurring. The maps look like simple type changes.

The changes that could occur here are:

- a disk's VCR changes
- an interface's opte port changes?
- ?

More comments would help.

hawkw replied:
I believe this is replacing the components that will be sent in the migration payload's instance spec by the migration source Propolis. We're telling it, "actually, you can't use the VCRs/OPTE ports that the source will tell you about, because they're for the previous incarnation of the instance, so use these VCRs/ports for those IDs instead". I agree that this should be documented here.

jmpesp commented on sled-agent/src/instance.rs:
```diff
@@ -867,64 +938,188 @@ impl InstanceRunner {
+            let migration_id = self
+                .state
+                .migration_in()
+                .expect("migration ensure request entails migration in")
```
Is this state something that's derived from the database, and therefore would this expect always fail due to some database state? Or is this state that is created a runtime? I'm worried this could cause a crash loop due to some database state.

hawkw replied:
The previous comment attempted to explain this a bit; it is part of the migration saga behavior and I believe things would have to have gone *very* wrong for this to ever happen. I think we had a discussion about it at some point during #5749 and felt that the expect was okay, but I haven't been able to actually track that down...

jmpesp commented on sled-agent/src/instance.rs:
```diff
@@ -867,64 +938,188 @@ impl InstanceRunner {
+            }
+
+            for DiskComponents { id, device, backend } in disks.into_iter() {
+                assert!(spec
```
Some of the inserts into `components` do not assert, some do - I'm wondering about the conditions being guarded against here.

jmpesp reviewed (approved):
I like it!

hawkw commented on nexus/db-queries/src/db/datastore/volume.rs:
```diff
@@ -2578,10 +2578,11 @@ fn region_in_vcr(
-                    if parsed_target == *region {
-                        region_found = true;
-                        break;
+                    if let SocketAddr::V6(target) = target {
```
yay for getting rid of `.parse()` calls!

hawkw commented on nexus/db-queries/src/db/datastore/volume.rs:
```diff
@@ -4435,28 +4442,31 @@ mod tests {
-        let address_1 = String::from("[fd00:1122:3344:104::1]:400");
-        let address_2 = String::from("[fd00:1122:3344:105::1]:401");
-        let address_3 = String::from("[fd00:1122:3344:106::1]:402");
+        let address_1: SocketAddrV6 =
+            "[fd00:1122:3344:104::1]:400".parse().unwrap();
+        let address_2: SocketAddrV6 =
+            "[fd00:1122:3344:105::1]:401".parse().unwrap();
+        let address_3: SocketAddrV6 =
+            "[fd00:1122:3344:106::1]:402".parse().unwrap();
```
wondered if it might be nicer to [programmatically construct these from an `Ipv6Addr` and port](https://doc.rust-lang.org/std/net/struct.SocketAddrV6.html#method.new) rather than parsing them, but honestly it's probably wordier than the current thing...

hawkw commented on nexus/src/app/background/tasks/region_snapshot_replacement_step.rs:
```diff
@@ -604,7 +604,7 @@ mod test {
                     opts: CrucibleOpts {
                         id: Uuid::new_v4(),
-                        target: vec![snapshot_addr],
+                        target: vec![snapshot_addr.parse().unwrap()],
```
nit, unimportant: could the test function just take `snapshot_addr` as a `SocketAddrV6`?

gjcolombo replied:
Possibly! I think this code was added after I originally drafted this change, and this was a mindless build fix after I rebased. I'll give that a shot.

gjcolombo replied:
Should be cleaned up in f0c692d.

hawkw commented on nexus/src/app/sagas/region_snapshot_replacement_step.rs:
```diff
@@ -555,11 +554,13 @@ async fn rsrss_notify_upstairs(
         "vmm id" => ?vmm.id,
     );
 
+    // N.B. The ID passed to this request must match the disk backend ID that
+    // sled agent supplies to Propolis when it creates the instance. Currently,
+    // sled agent uses the disk ID as the backend ID.
```
thanks for the comment!

hawkw reviewed (approved):
Overall, this looks good to me! I agree with @jmpesp that it would be nice to explain a few of the weird-looking bits.

gjcolombo reviewed (commented):
Thanks for the reviews, @jmpesp and @hawkw!

iximeow commented on nexus/tests/integration_tests/volume_management.rs:
```diff
@@ -3566,7 +3567,7 @@ struct TestReadOnlyRegionReferenceUsage {
 
     region: db::model::Region,
-    region_address: SocketAddrV6,
+    region_address: SocketAddr,
```
a great sadness washes over me

ok more seriously: this is because `CrucibleOpts::target`  is _also_ in terms of `SocketAddr` yeah? i'm not sweating test code nearly as much, just keeping a mental tally of what's pushing/pulling `CrucibleOpts::target`'s type.

iximeow commented on sled-agent/src/instance.rs:
```diff
@@ -867,64 +985,228 @@ impl InstanceRunner {
                         ),
+                    ))?;
+
+                let nic = &self.requested_nics[pos];
```
i know this is code motion, but the `iter/position/map(nics[pos])` could be `iter/find(nic.slot == port.slot())` without the index operation here (`pos` is otherwise unused and the `[]` made me think about panic safety...)

iximeow commented on sled-agent/src/instance.rs:
```diff
@@ -867,64 +985,228 @@ impl InstanceRunner {
+        // - Crucible disks: the target needs to connect to its downstairs
+        //   instances with new generation numbers supplied from Nexus
+        // - OPTE ports: the target needs to bind its VNICs to the correct
+        //   devices for its new host, which devices may have different names
```
```suggestion
        //   devices for its new host, which may have different names
```
extra "devices" i think? ("than" would reflow onto this line so if you agree then mind the rustfmt)

iximeow reviewed (approved):
yet more approve-with-comments, now with some type pedantics :) overall: really nice, thanks!

jmpesp commented on nexus/db-queries/src/db/datastore/volume.rs:
```diff
@@ -2578,10 +2578,11 @@ fn region_in_vcr(
+                            region_found = true;
+                            break;
+                        }
                     }
```
Mostly for testing on workstations / CI that aren't IPv6 machines

gjcolombo commented:
@iximeow Thanks for the review! I think I've addressed all your comments in f0c692d. I've also retested on a dev cluster that instance creation/start and disk snapshot seem to work (to try to gain some confidence that I didn't bungle any SocketAddr conversions or boot order designations).

I'm planning to merge this if/when CI comes back green.

gjcolombo commented:
CI failures are #7251, #7255, and #4657; requested a rerun.

gjcolombo commented:
Hmm, I see #7251 and #7255 should be fixed... looking a little more closely at those failures now (but still letting the rerun go).
